### ISSUE TITLE
Support Overlays

### ISSUE THREAD

**Issue by robotanarchy** [voidlinux/void-packages#3218]
[CONTRIBUTING.md](https://github.com/voidlinux/void-packages/blob/master/CONTRIBUTING.md#getting-your-packages-into-void-by-yourself) states:
> Please note that we do not accept any packages containing non-release versions such as specific git- or svn-revisions anymore.

So we need a feasible way to provide source repositories for these packages like the AUR or gentoo overlays. Providing a full fork of void-packages with >46K commits doesn't seem to be that great, because 99% of the repo would contain unrelated files and one would need to sync `common/shlibs` all the time.

I suggest overlay-repositories (folders), that look like this:
```
shlibs
custompackage1/
        template
custompackage2
        template
        patches/
                some-patch.patch
...
```

and adding a `--overlay` parameter to `xbps-src`:
`./xbps-src --overlay=~/code/void-packages-overlay pkg custompackage1`

**Comment by skrzyp** [comment]
Everyone can create own XBPS repositories, but I don't see that anyone do this…

**Comment by robotanarchy** [comment]
I will do this, as my [sway PR](https://github.com/voidlinux/void-packages/pull/3213) was declined because of the reason stated above.

**Comment by skrzyp** [comment]
But sway isn't complete yet.

**Comment by robotanarchy** [comment]
Yes, that is why my templates will not get included in void-packages. That's why I want to have this feature, to put my sway and wlc templates in an overlay.

**Comment by robotanarchy** [comment]
I realized by asking on IRC and testing, that xbps doesn't support changing package versions while building.
So for sway and wlc I will need a wrapper around xbps-src anyway, that sets the packages to the current git versions.
This means, that I don't really need this feature right now. I will post my wrapper script for xbps-src here, that will do basically the same thing as asked initially and support setting git-versions on the fly while packaging.

**Comment by Duncaen** [comment]
Templates are used for fixed versions of software, they should be reproducible.
Git based templates point to a specific commit to achieve that.

I like the idea of overlays but i see two problems at the moment, the `common/shlibs` file would be needed in overlays too. And how would you easily sync your overlayed packages with packages from `void-packages`, as example if you overwrite an existing template in your overlay to reduce the dependencies you can't just pull the latest changes. Maybe a script like `xbps-overlay` is needed to help maintaining overlays.

**Comment by robotanarchy** [comment]
Here's my repository, with an `overlay.sh` script that tries to do all that. It also merges `common/shlibs`.
https://github.com/robotanarchy/void-packages-bleeding-edge

**Comment by skrzyp** [comment]
It might be good to make this repo official, but less moderated, named `void-packages-wip` like NetBSD's pokgsrc does.

**Comment by robotanarchy** [comment]
@skrzyp: making this an official repo would be nice (just fork my repo into the voidlinux group then?), and I also like the '-wip' name, because it is shorter.

**Comment by Gottox** [comment]
What's the advantage of an overlay in contrast of using the power of git?

```
git fetch https://github.com/voidlinux/void-packages.git
git merge FETCH_HEAD
```

**Comment by robotanarchy** [comment]
Here are the features of my `overlay.sh`, I don't think you can all of that by just using git:
* No need to have the >46K commits in your own repository. The only really necessary file is the overlay.sh besides your partial `shlibs` file and `srcpkgs` folder.
* Caching all source code (git) repositories - the `void-packages` repo *and* all repos, that are used by the templates (`sway` and `wlc` repos in my case) (!). The repos only need to get updated, not completely downloaded again with all commits.
* `void-packages` repository gets fetched with the `--depth=1` git option, and then updated from there. So the backlog of >46K commits doesn't get downloaded at all.
* Templates in the `srcpkg` folder can have `version=LATEST` set. The overlay.sh script replaces that fake version with a version number generated from the git commit count and latest hash (like it is done in the AUR) after the merge.

I suggest reading the `overlay.sh` script and the two srcpkg-templates from my repo, it's actually very short :+1:

**Comment by Gottox** [comment]
Can you explain what this issue is about? Do you want us to ship overlay.sh? How would a workflow with it look like?

**Comment by robotanarchy** [comment]
This issue is generally about handling templates, that do not fit in the `void-packages` git repository.

I think shipping *something like* my `overlay.sh` would be useful; however, it won't work in the current state and properly integrating this in the main repo looks like a lot of effort to me.

The current workflow is documented [here](https://github.com/robotanarchy/void-packages-bleeding-edge#usage).

**Comment by ke7ofi** [comment]
Disclaimer: I don't yet understand the architecture, so this may not be technically feasible.

As far as I can tell from [the wiki](https://wiki.voidlinux.eu/XBPS#Repositories), it should be possible to create arbitrary remote repositories. What's stopping the creation of one designated for testing, similar to the “overlay” repo, which would have full integration (other than the initial step of adding it to `/etc/xbps.d`)? It might even be feasible to add a step in the installer (though I'm not sure how much you're willing to expand/bloat the installer) where users can say yes or no to a testing repo.

If it ends up being a simple (organisationally, if not technically) matter of forking @robotanarchy's bleeding-edge repo, I'd recommend transferring owners rather than forking it.

**Comment by ke7ofi** [comment]
Sorry, I think I misunderstood. The difference between this and the "overlay" repo is the ability to pull from VCS?

**Comment by robotanarchy** [comment]
@ke7ofi: The main difference is, that the repositories added to `/etc/xbps.d` are binary based, my overlay is for source packages only (because they need to be rebuilt daily or so).

Theoretically, official packages can pull from VCS (and there are still some packages, that do) too, but this isn't wanted anymore (see the very first post of this issue).

**Comment by ke7ofi** [comment]
What's the reason behind that policy, anyways?

**Comment by Duncaen** [comment]
There are some packages that are pulled from git, but having tarballs that can be cached, mirrored etc is nicer.

**Comment by ke7ofi** [comment]
That makes sense, thanks for the explanation.
